src/header-bidding/prebid/appnexus.spec.ts

```diff
@@ -0,0 +1,209 @@
+	containsMpuOrDmpu as containsMpuOrDmpu_,
+	getBreakpointKey as getBreakpointKey_,
+} from '../utils';
+import { getAppNexusDirectPlacementId } from './appnexus';
+
+jest.mock('../utils');
+const containsBillboardNotLeaderboard =
+	containsBillboardNotLeaderboard_ as jest.Mock;
+const containsDmpu = containsDmpu_ as jest.Mock;
+const containsLeaderboard = containsLeaderboard_ as jest.Mock;
+const containsLeaderboardOrBillboard =
+	containsLeaderboardOrBillboard_ as jest.Mock;
+const containsMobileSticky = containsMobileSticky_ as jest.Mock;
+const containsMpu = containsMpu_ as jest.Mock;
+const containsMpuOrDmpu = containsMpuOrDmpu_ as jest.Mock;
+const getBreakpointKey = getBreakpointKey_ as jest.Mock;
+
+jest.mock('utils/geo-utils');
+const isInAuOrNz = isInAuOrNz_ as jest.Mock;
+const isInRow = isInRow_ as jest.Mock;
+const isInUk = isInUk_ as jest.Mock;
+const isInUsa = isInUsa_ as jest.Mock;
+
+jest.mock('experiments/ab', () => ({
+	isInVariantSynchronous: jest.fn(),
+}));
+
+jest.mock('lib/cookies', () => ({
+	getCookie: jest.fn(),
+}));
+
+const resetConfig = () => {
+	window.guardian.ophan = {
+		pageViewId: 'pvid',
+		viewId: 'v_id',
+		record: () => {
+			// do nothing;
+		},
+		setEventEmitter: null,
+		trackComponentAttention: null,
+	};
+	window.guardian.config.switches = {
+		prebidAppnexus: true,
+		prebidAppnexusInvcode: true,
+		prebidOpenx: true,
+		prebidImproveDigital: true,
+		prebidIndexExchange: true,
+		prebidSonobi: true,
+		prebidTrustx: true,
+		prebidXaxis: true,
+		prebidAdYouLike: true,
+		prebidTriplelift: true,
+		prebidCriteo: true,
+	};
+	window.guardian.config.page.contentType = 'Article';
+	window.guardian.config.page.section = 'Magic';
+	window.guardian.config.page.isDev = false;
+};
+
+describe('getAppNexusDirectPlacementId', () => {
+	afterEach(() => {
+		jest.resetAllMocks();
+		resetConfig();
+	});
+
+	test('should return correct placementID for any existing biding sizes in Australia or New Zealand', () => {
+		isInAuOrNz.mockReturnValue(true);
+		getBreakpointKey.mockReturnValue('D');
+		containsMpu.mockReturnValue(true);
+		expect(getAppNexusDirectPlacementId([createAdSize(300, 250)])).toBe(
+			'11016434',
+		);
+	});
+
+	test('should return correct placementID for mobile-sticky and in ROW', () => {
+		isInRow.mockReturnValue(true);
+		getBreakpointKey.mockReturnValue('M');
+		containsMobileSticky.mockReturnValue(true);
+		expect(getAppNexusDirectPlacementId([createAdSize(320, 50)])).toBe(
+			'31512573',
+		);
+	});
+
+	test('should return correct placementID for ad sizes not in the conditional statement and not mobile-sticky in ROW', () => {
+		isInRow.mockReturnValue(true);
+		getBreakpointKey.mockReturnValue('D');
+		containsDmpu.mockReturnValue(true);
+		expect(getAppNexusDirectPlacementId([createAdSize(300, 600)])).toBe(
+			'9251752',
+		);
+	});
+
+	test('should return correct placementID for desktop billboard not leaderboad in UK, ROW and US', () => {
+		isInUk.mockReturnValue(true);
```

Comment:
I think here we should just mock `isInUk` and **not** `isInRow` or `isInUsa`. At the moment in this test we're telling the check that we're simultaneously in the UK, ROW and USA, which isn't a valid test case. This comment applies to all the following checks too, where we mock 3 location values at once, we should only mock 1.

Reply:
Ok after you explanation it totally makes sense! Done in 12648f9372f9fb09dbe7a793a47d6c6bdbe748aa